src/Microsoft.Graph.Core/Requests/BaseRequest.cs

```diff
@@ -278,6 +297,28 @@ private void AddHeadersToRequest(HttpRequestMessage request)
             }
         }
 
+        /// <summary>
+        /// Adds a <see cref="GraphRequestContext"/> to <see cref="HttpRequestMessage"/> property bag
+        /// </summary>
+        /// <param name="httpRequestMessage">A <see cref="HttpRequestMessage"/></param>
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/></param>
+        private void AddRequestContextToRequest(HttpRequestMessage httpRequestMessage, CancellationToken cancellationToken)
+        {
+            // Retreive feature flags
+            httpRequestMessage.Headers.TryGetValues(CoreConstants.Headers.FeatureFlag, out IEnumerable<string> featureFlags);
+
+            // Creates a request context object
+            var requestContext = new GraphRequestContext
+            {
+                MiddlewareOptions = MiddlewareOptions,
+                ClientRequestId = GetHeaderValue(httpRequestMessage, CoreConstants.Headers.ClientRequestId) ?? Guid.NewGuid().ToString(),
```

Comment:
Now that we can add the ClientRequestId header via the RequestContext, we should update AddHeadersToRequest to overwrite an existing ClientRequestId with the one provided by the dev. This is only an issue in case we change the order of `AddHeadersToRequest` and `AddRequestContextToRequest` method calls in `GetHttpRequestMessage()`. We want the customer supplied ClientRequestId to always override an existing ClientRequestId.

Reply:
I have added a check in `GetHeaderValue` to handle this scenario.